content/advent-2016/http-client.md

```diff
@@ -0,0 +1,272 @@
+possible, it's good to try to be forward compatible in your client library. For
+example, if the server offers a new parameter, or changes the available types
+for an existing parameter, users should be able to specify those without
+needing to pull down the latest versions.
+
+Specifying API parameters with a `url.Values` works really well for this use
+case. For example:
+
+```go
+data := url.Values{}
+data.Set("To", "+14105551234")
+data.Set("From", "+14105556789")
+client.Calls.Create(data)
+```
+
+If the server decides to allow calls to people's names instead of phone
+numbers, or to the number 7, or allow multiple From values, or a new parameter,
+your users are totally compatible with their existing code! They can just
+change the values they set on `data` and they are good to go.
+
+## Usage Patterns
+
+For most client libraries, the vast majority of your users will only do one
+or two things with the API. For Stripe this is charging a credit card, for
+Sendgrid this is sending an email, &c, &c. Offer helpers to make common actions
+really easy. For example, this type of interface is easy to scale to many
+resources and many different HTTP methods:
+
+```go
+data := url.Values{}
+data.Set("From", "boss@example.org")
+data.Set("To", "foo@example.com")
+data.Set("Subject", "TPS Cover Sheets")
+data.Set("Body", "They are important!")
+return client.V1.Emails.CreateResource(data)
+```
+
+But it's a little cumbersome. It might be worthwhile to add a helper function
+that simplifies the interface a little bit.
+
+```go
+client.SendEmail("boss@example.org", "foo@example.com", "TPS Cover Sheets",
+    "They are important! Don't forget them.")
+```
+
+## Testing
+
+One way to test your client would be to define each resource as an interface,
+then add dummy code that satisfies the interface and returns you objects. For
+example:
+
+```go
+type ChargeResource interface {
+    Get(string) (*Charge, error)
+    Create(url.Values) (*Charge, error)
+    List(url.Values) ([]*Charge, error)
+}
+```
+
+Then your tests would integrate with dummy versions of each interface. The
+problem here is that you're not actually testing against the HTTP response. If
+you make a change to the client code that parses the HTTP response incorrectly,
+you're not going to catch it.
+
+The next option is to integrate directly with the API - pass in a valid set of
+credentials and make network requests. This is the only way to tell if the API
+starts returning different responses, but is slow, doesn't work on the subway,
+and may be expensive, if you are testing API calls that cost money.
+
+The third option is to save the API response, then spin up a test server
+that serves that response on demand. This gets you most of the benefits of
+integration with the API, but is much faster, works on the subway and won't
+cost you anything to test expensive calls. Spinning up a test server is easy
+and cheap in Go. Here's an example test:
+
+```
```

Comment:
The other code snippets have included `go` here, was this intentional or just missed?